core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -816,6 +816,7 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
     if (reassignment.isEmpty) {
       info("No more partitions need to be reassigned. Deleting zk path %s".format(ReassignPartitionsZNode.path))
       zkClient.deletePartitionReassignment()
+      zkClient.registerZNodeChangeHandlerAndCheckExistence(partitionReassignmentHandler)
```

Comment:
@onurkaraman, @junrao There is a window where the reassignment znode could be created after the deletion but before this call. Does that mean that we would not detect such a reassignment? If that's the case, what's the best way to handle it?

Reply:
@ijuma: This window is why I stated the following earlier:
> we could explore basically calling `KafkaZkClient.registerZNodeChangeHandlerAndCheckExistence` at the end of rebalance and maybe enqueuing a PartitionReassignment if something exists.
> 
> We probably would want to change `KafkaZkClient.registerZNodeChangeHandlerAndCheckExistence` to return true or false.

To rephrase my earlier comment, we want to not only register the watcher, but also see if the znode exists. If the znode exists after our deletion, then we should enqueue PartitionReassignment event. Luckily for us, ExistsRequest lets us do both in one request.

Your change to call `KafkaZkClient.registerZNodeChangeHandlerAndCheckExistence` after deletion already does most of the work: it sends out an ExistsRequest while registering the watcher in the same request. The only problem is that we're currently ignoring the existence result.

In order to make this work, we want to change `KafkaZkClient.registerZNodeChangeHandlerAndCheckExistence` to return true or false so that we can enqueue a PartitionReassignment event if true.

More specifically, `KafkaZkClient.registerZNodeChangeHandlerAndCheckExistence` should end up looking like this:
```
if (existsResponse.resultCode == Code.OK) true
else if (existsResponse.resultCode == Code.NONODE) false
else throw existsResponse.resultException.get
```
Once we make that change, then we'd just change the code to the following:
```
if (zkClient.registerZNodeChangeHandlerAndCheckExistence(partitionReassignmentHandler)) {
  eventManager.put(controller.PartitionReassignment)
}
```
Doing so will cover any reassignment znode creation made during this window.

Reply:
Thanks, my bad for missing what you meant originally.